codegen/layout_tool.py

```diff
@@ -0,0 +1,314 @@
+        if check_single(p, ent2_points):
+            return True
+    for p in ent2_points:
+        if check_single(p, ent1_points):
+            return True
+    return False
+
+# this is a pretty ugly hack.
+# we want this callback to be a method of LayoutTool, so that no global variables can be used
+# however, being a method of a class means that 
+# 1 - first argument is this class
+# 2 - it is somehow registered in class __dict__ field
+# since opencv doesn't really allow to modify it's callback procedure, we will create this funciton
+# the signature is of standard callback, with class_ref added.
+# in collect.py/launch_layout_tool() you can see that this function is patched with LayoutTool reference
+# meaning that this behaves _exactly_ like a method of LayoutTool with two exceptions:
+# - you cannot call layout_tool_instance.callback()
+# - instead of self. you need to use class_ref.
+def callback(class_ref, event,x,y,flags,param): 
+    if event == cv2.EVENT_LBUTTONDOWN:
+        if class_ref.pointerbound:
+            logging.info(f"dropped cell {class_ref.pointerbound}")
+            class_ref.pointerbound = None
+            class_ref.update_image()
+        else:
+            for name, data in class_ref.layout_info.items():
+                start = data["design_location"]
+                end = [
+                    data["design_location"][0] + data["design_area"][0],
+                    data["design_location"][1] + data["design_area"][1]
+                ]
+
+                if x > start[0] and x < end[0]:
+                    if y > start[1] and y < end[1]:
+                        if data["grabbable"]:
+                            class_ref.pointerbound = name
+                            logging.info(f"picked cell {class_ref.pointerbound}")
+
+    elif event == cv2.EVENT_MOUSEMOVE:
+        if class_ref.pointerbound:
+            layout_state = "good"
+
+            class_ref.layout_info[class_ref.pointerbound]["design_location"] = (int(x), int(y))
+            class_ref.layout_info[class_ref.pointerbound + "__border"]["design_location"] = (
+                int(x - class_ref.border_offsets // class_ref.scalar), 
+                int(y - class_ref.border_offsets // class_ref.scalar)
+            )
+            
+            collision_map = {}
+            for e in class_ref.layout_info.keys():
+                collision_map[e] = False
+
+            for ent_name, data in class_ref.layout_info.items():
+                    for o_name, odata in class_ref.layout_info.items():
+                            if ent_name + "__border" != o_name and ent_name != o_name + "__border":
+                                if ent_name != o_name:
+                                    if check_collision(data, odata):
+                                        collision_map[ent_name] = True
+                                        collision_map[o_name] = True
+            
+            for ent_name, data in class_ref.layout_info.items():
```

Comment:
It might make sense to use an enum for cell states (good/violating/bad) but I decided to go with assigning the color and checking against it. Macros wouldn't really add much value here (imo) and wouldn't really shrink the code.

`layout_state` seems to be unused, I decided to axe it's user for now. tldr on it, is that it contains the worst offender of the entire design.